app/src/main/java/com/wednesday/banklocator/adapter/IfscAdapter.kt

```diff
@@ -0,0 +1,55 @@
+package com.wednesday.banklocator.adapter
+
+import android.view.Gravity
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+import android.widget.Toast
+import androidx.recyclerview.widget.RecyclerView
+import com.wednesday.banklocator.R
+import com.wednesday.banklocator.model.IfscResponse
+import com.wednesday.banklocator.viewmodel.BankViewModel
+import kotlinx.android.synthetic.main.partial_banks_list.view.*
+
+class IfscAdapter(var item: ArrayList<IfscResponse>, private val viewModel: BankViewModel) :
+    RecyclerView.Adapter<IfscAdapter.bankDetailsViewHolder>(
+    ) {
+    inner class bankDetailsViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView)
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): bankDetailsViewHolder {
+        return bankDetailsViewHolder(
+            LayoutInflater.from(parent.context).inflate(
+                R.layout.partial_banks_list,
+                parent,
+                false
+            )
+        )
+    }
+
+    override fun onBindViewHolder(holder: bankDetailsViewHolder, position: Int) {
+        val model = item
+        holder.itemView.apply {
+            bankNameTextView.text = model.BRANCH
+            bankAddressTextView.text = model.ADDRESS
+            favouritesIconImageView.setOnClickListener {
+                val toast = Toast.makeText(context, model.BRANCH+" added to Local Database", Toast.LENGTH_LONG)
+                toast.setGravity(Gravity.CENTER_VERTICAL,0,0)
+                toast.show()
+                viewModel.upsert(model)
```

Comment:
I don't like this. Basically, when you pass a model down to an adapter or any other object to an adapter the object is doing more than what it's supposed to do. The adapter's job is to get a list of objects and map it to the views in a list, recycler view, etc. That's it. Anything more and it's difficult to test.

Here you should use interfaces and pass a callback when the favorite button is pressed. That way you don't have a tight coupling with the bank view model.

Reply:
Okay

Reply:
![Alden-screen-recording](https://user-images.githubusercontent.com/59332509/104152793-9a49d900-5406-11eb-92b8-300f099709a7.gif)

Reply:
![ezgif com-gif-maker](https://user-images.githubusercontent.com/59332509/104153053-55727200-5407-11eb-94fa-1bd4051727df.gif)